### PR TITLE
(PUP-6768) Add Pcore Object type Puppet::LookupContext

### PR DESCRIPTION
**Pull request by thallgren**
This PR adds a new Pcore Object type named `Puppet::LookupContext`
with attributes and methods.

### CODE REVIEW

File: lib/puppet/pops/lookup/context.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            }
+          },
+          {
+            'not_found' => tf.callable([0, 0], tf.undef),
+            'explain' => tf.callable([0, 0, tf.callable(0,0)], tf.undef),
+            'cache' => tf.callable([tf.scalar, tf.any], tf.undef),
+            'cache_all' => tf.callable([tf.hash_kv(tf.scalar, tf.any)], tf.undef),
+            'cached_value' => tf.callable([tf.scalar], tf.any),
+            'cached_entries' => tf.variant(
+              tf.callable([0, 0, tf.callable(2,2)], tf.undef),
```
Comment (hlindberg) [inline lib/puppet/pops/lookup/context.rb]: If it is to mimic each, then also support 1 arg in the lambda to get each key/value tuple ?

File: spec/lib/puppet_spec/unindent.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class String
+  def unindent
```
Comment (hlindberg) [inline spec/lib/puppet_spec/unindent.rb]: isn't there a risk in monkey patching String this way when running tests?

File: spec/unit/pops/lookup/context_spec.rb [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    it 'can use the cache to merge hashes and individual entries' do
+      code = <<-PUPPET.unindent
+        $ctx = Puppet::LookupContext.new('e', 'm')
+        $ctx.cache_all({ 'v1' => 'first', 'v2' => 'second' })
+        $ctx.cache('v3', 'third')
+        $ctx.cached_entries |$key, $value| { notice($key); notice($value) }
+      PUPPET
+      expect(eval_and_collect_notices(code)).to eql(%w(v1 first v2 second v3 third))
+    end
+
```
Comment (hlindberg) [inline spec/unit/pops/lookup/context_spec.rb]: test missing: can replace a value for a cache key `cache('x', 10); notice cached_value('x'); cache('x', 20); notice cached_value('x');`

File: spec/unit/pops/lookup/context_spec.rb [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        $ctx.not_found
+      PUPPET
+      expect { eval_and_collect_notices(code) }.to throw_symbol(:no_such_key)
+    end
+
+    context 'when used in an Invocation' do
+      let(:invocation) { Invocation.new({}) }
+      let(:invocation_with_explain) { Invocation.new({}, {}, {}, true) }
+
+      it 'will no call explain unless explanations are active' do
```
Comment (hlindberg) [inline spec/unit/pops/lookup/context_spec.rb]: typo "will no call"

File: spec/lib/puppet_spec/unindent.rb (hunk 2 above)
Comment (thallgren) [inline spec/lib/puppet_spec/unindent.rb]: If someone else also monkey patches String with the same method and does something differently, then a lot of tests will fail of course, but I'm not sure I'd call that a risk.

**Pull request merged**